// @akord/akord-js/lib/types/membership.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Membership = void 0;
const crypto_1 = require("@akord/crypto");
const profile_details_1 = require("./profile-details");
class Membership extends crypto_1.Encryptable {
    constructor(membershipProto, keys) {
        var _a, _b, _c, _d;
        super(keys, null);
        this.id = membershipProto.id;
        this.owner = membershipProto.owner;
        this.createdAt = membershipProto.createdAt;
        this.updatedAt = membershipProto.updatedAt;
        this.data = membershipProto.data;
        this.status = membershipProto.status;
        this.role = membershipProto.role;
        this.encPublicSigningKey = membershipProto.encPublicSigningKey;
        this.vaultId = membershipProto.vaultId;
        this.keys = keys;
        this.memberDetails = new profile_details_1.ProfileDetails((_a = membershipProto.memberDetails) === null || _a === void 0 ? void 0 : _a.name, (_b = membershipProto.memberDetails) === null || _b === void 0 ? void 0 : _b.publicSigningKey, (_c = membershipProto.memberDetails) === null || _c === void 0 ? void 0 : _c.email, (_d = membershipProto.memberDetails) === null || _d === void 0 ? void 0 : _d.avatarUri, keys, null);
    }
}
exports.Membership = Membership;
