// @akord/akord-js/lib/types/node.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.StorageType = exports.NodeFactory = exports.MemoReaction = exports.MemoVersion = exports.FileVersion = exports.Memo = exports.Note = exports.Stack = exports.Folder = exports.Node = void 0;
const crypto_1 = require("@akord/crypto");
class Node extends crypto_1.Encryptable {
    constructor(id, createdAt, updatedAt, status, vaultId, owner, data, parentId, keys, publicKey) {
        super(keys, publicKey);
        this.id = id;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
        this.status = status;
        this.vaultId = vaultId;
        this.owner = owner;
        this.data = data;
        this.parentId = parentId;
    }
}
exports.Node = Node;
class Folder extends Node {
    constructor(nodeLike, keys) {
        super(nodeLike.id, nodeLike.createdAt, nodeLike.updatedAt, nodeLike.status, nodeLike.vaultId, nodeLike.owner, nodeLike.data, nodeLike.parentId, keys);
        this.name = nodeLike.name;
    }
}
__decorate([
    (0, crypto_1.encrypted)()
], Folder.prototype, "name", void 0);
exports.Folder = Folder;
class Stack extends Node {
    constructor(nodeLike, keys) {
        super(nodeLike.id, nodeLike.createdAt, nodeLike.updatedAt, nodeLike.status, nodeLike.vaultId, nodeLike.owner, nodeLike.data, nodeLike.parentId, keys);
        this.name = nodeLike.name;
        this.versions = (nodeLike.versions || []).map((version) => new FileVersion(version, keys));
    }
}
__decorate([
    (0, crypto_1.encrypted)()
], Stack.prototype, "name", void 0);
exports.Stack = Stack;
class Note extends Node {
    constructor(nodeLike, keys) {
        super(nodeLike.id, nodeLike.createdAt, nodeLike.updatedAt, nodeLike.status, nodeLike.vaultId, nodeLike.owner, nodeLike.data, nodeLike.parentId, keys);
        this.name = nodeLike.name;
        this.versions = (nodeLike.versions || []).map((version) => new FileVersion(version, keys));
    }
}
__decorate([
    (0, crypto_1.encrypted)()
], Note.prototype, "name", void 0);
exports.Note = Note;
class Memo extends Node {
    constructor(nodeLike, keys, publicKey) {
        super(nodeLike.id, nodeLike.createdAt, nodeLike.updatedAt, nodeLike.status, nodeLike.vaultId, nodeLike.owner, nodeLike.data, nodeLike.parentId, keys, publicKey);
        this.versions = (nodeLike.versions || []).map((version) => new MemoVersion(version, keys, publicKey));
    }
}
exports.Memo = Memo;
class FileVersion extends crypto_1.Encryptable {
    constructor(fileVersionProto, keys, publicKey) {
        super(keys, publicKey);
        this.owner = fileVersionProto.owner;
        this.type = fileVersionProto.type;
        this.resourceUri = fileVersionProto.resourceUri;
        this.createdAt = fileVersionProto.createdAt;
        this.size = fileVersionProto.size;
        this.numberOfChunks = fileVersionProto.numberOfChunks;
        this.chunkSize = fileVersionProto.chunkSize;
        this.name = fileVersionProto.name;
        this.status = fileVersionProto.status;
    }
    getUri(type) {
        var _a, _b;
        return (_b = (_a = this.resourceUri) === null || _a === void 0 ? void 0 : _a.find(uri => uri.startsWith(type))) === null || _b === void 0 ? void 0 : _b.replace(type, "");
    }
}
__decorate([
    (0, crypto_1.encrypted)()
], FileVersion.prototype, "name", void 0);
exports.FileVersion = FileVersion;
class MemoVersion extends crypto_1.Encryptable {
    constructor(memoVersionProto, keys, publicKey) {
        super(keys, publicKey);
        this.owner = memoVersionProto.owner;
        this.createdAt = memoVersionProto.createdAt;
        this.message = memoVersionProto.message;
        this.reactions = (memoVersionProto.reactions || []).map((reaction) => new MemoReaction(reaction, keys, publicKey));
        this.attachments = (memoVersionProto.attachments || []).map((attachment) => new FileVersion(attachment, keys));
    }
}
__decorate([
    (0, crypto_1.encrypted)()
], MemoVersion.prototype, "message", void 0);
exports.MemoVersion = MemoVersion;
class MemoReaction extends crypto_1.Encryptable {
    constructor(memoReactionProto, keys, publicKey) {
        super(keys, publicKey);
        this.owner = memoReactionProto.owner;
        this.createdAt = memoReactionProto.createdAt;
        this.reaction = memoReactionProto.reaction;
    }
}
__decorate([
    (0, crypto_1.encrypted)()
], MemoReaction.prototype, "reaction", void 0);
exports.MemoReaction = MemoReaction;
class NodeFactory {
    static instance(nodeLike, data, keys) {
        return new nodeLike(data, keys);
    }
}
exports.NodeFactory = NodeFactory;
var StorageType;
(function (StorageType) {
    StorageType["S3"] = "s3:";
    StorageType["ARWEAVE"] = "arweave:";
})(StorageType = exports.StorageType || (exports.StorageType = {}));
