// @akord/akord-js/lib/types/vault.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Vault = void 0;
const crypto_1 = require("@akord/crypto");
class Vault extends crypto_1.Encryptable {
    constructor(vaultProto, keys) {
        var _a;
        super(keys, null);
        this.id = vaultProto.id;
        this.public = vaultProto.public;
        this.createdAt = vaultProto.createdAt;
        this.updatedAt = vaultProto.updatedAt;
        this.size = (_a = vaultProto.storage) === null || _a === void 0 ? void 0 : _a.storage_used;
        this.name = vaultProto.name;
        this.status = vaultProto.status;
        this.data = vaultProto.data;
        this.keys = keys;
    }
}
__decorate([
    (0, crypto_1.encrypted)()
], Vault.prototype, "name", void 0);
exports.Vault = Vault;
